utils.py

```diff
@@ -0,0 +1,179 @@
+        assert remaining_ports, "Service ports are exhausted"
+
+        # We don't want to be searching forever, so limit the number of times
+        # we clash to the number of remaining ports.
+        ports = self.ports_by_app.values()
+        port = None
+        for i in range(MAX_CLASHES):
+            hash_str = "%s-%s-%s" % (app['id'], task_port, i)
+            hash_val = hashlib.sha1(hash_str.encode("utf-8")).hexdigest()
+            hash_int = int(hash_val[:8], 16)
+            trial_port = self.min_port + (hash_int % self.max_ports)
+            if trial_port not in ports:
+                port = trial_port
+                break
+        if port is None:
+            for port in range(self.min_port, self.max_port + 1):
+                if port not in ports:
+                    break
+
+        # We must have assigned a unique port by now since we know there were
+        # some available.
+        assert port and port not in ports
+
+        logger.debug("Assigned new port: %d", port)
+        return port
+
+    def _get_service_port(self, app, task_port):
+        key = (app['id'], task_port)
+        port = (self.ports_by_app.get(key) or
+                self._assign_new_service_port(app, task_port))
+        self.ports_by_app[key] = port
+        return port
+
+    def set_ports(self, min_port, max_port):
+        """
+        Set the range of ports that we can use for auto-assignment of
+        service ports - just for IP-per-task apps.
+        :param min_port: The minimum port value
+        :param max_port: The maximum port value
+        """
+        self.min_port = min_port
+        self.max_port = max_port
+        self.max_ports = max_port - min_port + 1
+        self.can_assign = self.min_port and self.max_port
+        assert not self.ports_by_app
+        assert self.max_ports > 1
+
+    def reset(self):
+        """
+        Reset the assigner so that ports are newly assigned.
+        :return:
+        """
+        self.ports_by_app = {}
+
+    def get_service_ports(self, app):
+        ports = app['ports']
+        if not ports and is_ip_per_task(app) and self.can_assign:
+            logger.warning("Auto assigning service port for "
+                           "IP-per-container task")
+            task = app['tasks'][0]
+            _, task_ports = get_task_ip_and_ports(app, task)
+            ports = [self._get_service_port(app, task_port)
+                     for task_port in task_ports]
+        logger.debug("Service ports: %r", ports)
+        return ports
+
+
+def resolve_ip(host):
+    cached_ip = ip_cache.get(host, None)
+    if cached_ip:
+        return cached_ip
+    else:
+        try:
+            logger.debug("trying to resolve ip address for host %s", host)
+            ip = socket.gethostbyname(host)
+            ip_cache[host] = ip
+            return ip
+        except socket.gaierror:
+            return None
+ip_cache = dict()
```

Comment:
It's occurred to me that this should be an LRU cache, otherwise it'll leak. I can fix this in a follow-up PR.